### ISSUE TITLE
Question about "Unit Action Simulation (UAS)"  in "ppo_diverse.py"

### ISSUE THREAD

**Issue by ZJJ2017** [vwxyzjn/gym-microrts-paper#2]
I have a question about the way of issuing actions to a number of units——UAS. The paper "Gym-URTS: Toward Affordable Full Game Real-time Strategy Games Research with Deep Reinforcement Learning" wrote that "Unit Action Simulation (UAS) calls the policy iteratively until all units have actions (at each step, the policy chooses a unit and issue a unit action to it)." In the "ppo_diverse.py", each frame, the policy only  chooses one unit and issue a action to it, which can not realize the function of issuing actions to a number of units at one frame. Relative code are listed as follow.

source_unit_mask = torch.Tensor(np.array(envs.vec_client.getUnitLocationMasks()).reshape(args.num_envs, -1))
multi_categoricals = [CategoricalMasked(logits=split_logits[0], masks=source_unit_mask)]
action_components = [multi_categoricals[0].sample()]
source_unit_action_mask = torch.Tensor(  np.array(envs.vec_client.getUnitActionMasks(action_components[0].cpu().numpy())).reshape(args.num_envs, -1))
split_suam = torch.split(source_unit_action_mask, envs.action_space.nvec.tolist()[1:], dim=1)
multi_categoricals = multi_categoricals + [CategoricalMasked(logits=logits, masks=iam) for (logits, iam) in zip(split_logits[1:], split_suam)]
invalid_action_masks = torch.cat((source_unit_mask, source_unit_action_mask), 1)
action_components += [categorical.sample() for categorical in multi_categoricals[1:]]
action = torch.stack(action_components)

**Comment by vwxyzjn** [comment]
Thank you for your interest. The way it works is a bit hacky hence we have deprecated UAS in favor of Gridnet in future work. But to answer your question, the “calls policy iteratively” part happens in the environment, hence you are not seeing the code in ppo_diverse.py.

Specifically, the code is here https://github.com/vwxyzjn/microrts/blob/3b0d5b60777f0824f10cc6f13a7777117212fccb/src/tests/JNIClient.java#L135 where if not all units have been issued an action we simulate a fake game state, and only issue action to the real games sate until all units have been issued an action.
